### ISSUE TITLE
fails to compile large modules on ts 3.5.3

### ISSUE THREAD

**Issue by Toxicable** [bazelbuild/rules_nodejs#961]
# 🐞 bug report

### Affected Rule
`ts_library`

### Is this a regression?
Yes, works fine on typescript@3.4.x

### Description
Compiling a largish target which on 3.4.x would take ~8 seconds, but only 3.5.x it takes ~60s but never completes since it hits an eventual out of memory error

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>linux x64</code>
</pre>

**Output of `bazel version`:**

<pre>
  <code>Build label: 0.27.0
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Mon Jun 17 13:00:46 2019 (1560776446)
Build timestamp: 1560776446
Build timestamp as int: 1560776446</code>
</pre>

**Rules version (SHA):**

<pre>
  <code>0.35.0</code>
</pre>

**Anything else relevant?**

**Comment by alexeagle** [comment]
TypeScript itself had a big performance regression in 3.4 but it was fixed in 3.5.
Is this reproducible outside of Bazel with just `tsc`? Can you pare down a repro?

In google we found some big perf regressions in particular code patterns, where a single file could demonstrate the problem. I don't know of anything like this for TS 3.5 - @evmar is doing that upgrade in Google right now so maybe he has also seen some perf regressions?

**Comment by evmar** [comment]
I haven't seen this.  I have seen TS3.5 reject programs that 3.4 accepted due to overcomplex types, but it does so with a compiler error saying as much.  I would suggest trying to reduce the failure a bit.

**Comment by Toxicable** [comment]
>Is this reproducible outside of Bazel with just tsc

I modified our tsconfig.json to have `files: ["./path/to/module/main.ts"]` and ran `yarn tsc -p tsconfig.json`. I also set `export NODE_OPTIONS=--max_old_space_size=16345`

On 3.4.x it took 7s and on 3.5.x it's been going for a few mintes now and hasn't finished.
Going to take a look around on the TypeScript repository to see if anyone else has similar issues.
